### ISSUE TITLE
SAMBA auth does not work

### ISSUE THREAD

**Issue by NdK73** [BeyondTrust/pbis-open#88]
**Version**: 8.5.6.375 from pbis-open repository
**OS/Distro**: Debian Stretch
**Issue/Impact**: Samba shares for domain users are inaccessible; public shares are OK.

The system was operational with Samba (3.3.6) + Winbind. I upgraded it to Stretch (that includes Samba 4.5.12) and modified /etc/samba/smb.conf according to the integration guide.

```
root@Str957-share:~# net ads testjoin
Join is OK
```

In the log I get:
```
[2017/12/12 13:53:51.944205,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[homes]"
[2017/12/12 13:53:51.944847,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[for_web$]"
[2017/12/12 13:53:51.945184,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[web_users]"
[2017/12/12 13:53:51.945417,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[ArchivioStorico]"
[2017/12/12 13:53:51.945667,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[Software]"
[2017/12/12 13:53:51.945912,  2] ../source3/param/loadparm.c:2685(lp_do_section)
  Processing section "[ISO]"
[2017/12/12 13:53:51.964487,  2] ../source3/auth/auth.c:305(auth_check_ntlm_password)
  check_ntlm_password:  authentication for user [domain.user] -> [domain.user] -> [domain.user] succeeded
```
then nothing and smbclient answers with "session setup failed: NT_STATUS_IO_TIMEOUT".
If I increase the log level, I can see:
```
[2017/12/12 12:35:58.308148,  2, pid=25900, effective(0, 0), real(0, 0), class=auth] ../source3/auth/auth.c:305(auth_check_ntlm_password)
  check_ntlm_password:  authentication for user [domain.user] -> [domain.user] -> [domain.user] succeeded
[2017/12/12 12:35:58.308160, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/auth/auth_ntlmssp.c:215(auth3_check_password)
  Got NT session key of length 16
[2017/12/12 12:35:58.308170, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/auth/auth_ntlmssp.c:222(auth3_check_password)
  Got LM session key of length 8
[2017/12/12 12:35:58.308407, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/auth/token_util.c:224(create_local_nt_token_from_info3)
  Create local NT token for domain.user
[2017/12/12 12:35:58.308639, 10, pid=25900, effective(0, 0), real(0, 0), class=tdb] ../source3/lib/gencache.c:326(gencache_set_data_blob)
  Adding cache entry with key=[IDMAP/SID2XID/S-1-5-21-2162351890-1506888927-3107636301-8536] and timeout=[gio gen  1 01:00:00 1970 CET] (-1513078558 seconds in the past)
[2017/12/12 12:35:58.309142, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/passdb/lookup_sid.c:1468(sid_to_uid)
  sid S-1-5-21-2162351890-1506888927-3107636301-8536 -> uid 2125996376
[2017/12/12 12:35:58.309667, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/lib/system_smbd.c:176(sys_getgrouplist)
  sys_getgrouplist: user [domain.user]
[2017/12/12 12:35:58.320411, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/lib/system_smbd.c:176(sys_getgrouplist)
  sys_getgrouplist: user [domain.user]
[2017/12/12 12:35:58.326711, 10, pid=25900, effective(0, 0), real(0, 0), class=tdb] ../source3/lib/gencache.c:326(gencache_set_data_blob)
  Adding cache entry with key=[IDMAP/GID2SID/2125988353] and timeout=[gio gen  1 01:00:00 1970 CET] (-1513078558 seconds in the past)
[2017/12/12 12:35:58.327804, 10, pid=25900, effective(0, 0), real(0, 0)] ../source3/passdb/lookup_sid.c:1300(gid_to_sid)
  gid 2125988353 -> sid S-1-5-21-2162351890-1506888927-3107636301-513
[2017/12/12 12:35:58.327855, 10, pid=25900, effective(0, 0), real(0, 0), class=tdb] ../source3/lib/gencache.c:326(gencache_set_data_blob)
  Adding cache entry with key=[IDMAP/GID2SID/1906845086] and timeout=[gio gen  1 01:00:00 1970 CET] (-1513078558 seconds in the past)
```
[a lot more of gid->sid mappings and cache settings with timeouts in the past].

I noticed that restarting smbd became really slow (takes 1'30") if I have attempted to access an authenticated share with a domain user, but is immediate if I don't try to access an authenticated share with a domain user (a local user is OK). Possibly it's waiting for some child to timeout?

I tried starting / stopping winbindd and nmbd, but the result is always the same.

Can someone spot what I'm doing wrong?

Tks.

**Comment by rbest-bt** [comment]
Did you issue the samba-interop-install again after the system update?

**Comment by NdK73** [comment]
Yep. Sure. Before verifying "net ads join".

**Comment by rbest-bt** [comment]
How many users are there in your domain? It seems that authentication is working but it could be hanging up on enumerating membership. Name service cache daemon (nscd) on the host could help with this.

**Comment by NdK73** [comment]
Enumerating the users on our domain is a really stupid thing to do: more than 400k users and 600k groups *just in one of the domains* (the one the machine is joined to). Another domain in the forest is about 5 times bigger...
The first (and only) time I used "enum users" the server hogged the net for a day and a half... :(
In smb.conf "enum users" is obviously off..

**Comment by NdK73** [comment]
That would be useless for me: I have no access to the AD side except for creating machine accounts and groups (I'm not an AD admin, I can only manage a single OU).
But why does it need to enumerate users? Maybe it's an "optimization" that makes small domains faster, but renders bigger ones unusable. What's the workflow used for the authentication for requests received via Samba? Why is it different from users login (that works perfectly most of the time)?

**Comment by NdK73** [comment]
I made it work. Seems PBIS is not doing anything stupid like trying to enumerate all users in the domain :)

I ran:
```
/opt/pbis/bin/samba-interop-install --uninstall
apt uninstall pbis-open pbis-open-upgrade
rm -rf /var/lib/pbis
apt install pbis-open
/opt/pbis/bin/domainjoin-cli join my.domain my.user
/opt/pbis/bin/samba-interop-install --install
shutdown -r now
```
After reboot, running `id MY.DOMAIN\\my.user` takes 63ms instead of ~1'20" that it took before.
So it seems the upgrade triggered some corner case that leads to a drastic slowdown, with a lsass process taking 100% of the CPU for a long time.
Moreover, before doing this uninst/clear/reinst it took about 10-15 **minutes** before lsass started returning user info after every reboot (or service lwsmd restart).
